aten/src/ATen/native/sparse/SparseFactories.cpp

```diff
@@ -0,0 +1,176 @@
+  // forward.
+  Tensor indices = at::empty({2, max_nnz}, indices_options);
+  Tensor values = at::empty({max_nnz}, diagonals.options());
+
+  // Get accessors on input, and output
+  auto indices_accessor = indices.accessor<int64_t, 2>();
+  auto offsets_accessor = offsets.accessor<int64_t, 1>();
+  int64_t actual_nnz = 0;
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(
+      at::ScalarType::BFloat16,
+      at::ScalarType::Half,
+      at::ScalarType::Bool,
+      at::ScalarType::ComplexHalf,
+      values.scalar_type(),
+      "spdiags",
+      [&] {
+        auto values_accessor = values.accessor<scalar_t, 1>();
+        auto const diagonals_accessor = diagonals.accessor<scalar_t, 2>();
+
+        for (const auto d_i : c10::irange(n_diag)) {
+          const int64_t& off_i = offsets_accessor[d_i];
+          const int64_t row_out_begin = off_i < 0 ? std::abs(off_i) : 0;
+          const int64_t col_out_begin = off_i > 0 ? off_i : 0;
+
+          // Number of column, row positions we can assign into and the number
+          // of diag positions we can read from
+          const int64_t row_slots = n_rows_out - row_out_begin;
+          const int64_t col_slots = n_cols_out - col_out_begin;
+          const int64_t diag_slots = n_cols_in - col_out_begin;
+
+          // Max number of reads lets us limit the loop by the first thing we
+          // will exhaust, note if any of these comes up negative irange will be
+          // empty so we don't have to handle that case explicitly
+          const int64_t max_read =
+              std::min(diag_slots, std::min(row_slots, col_slots));
+
+
+          for (const auto read_count : c10::irange(max_read)) {
+            const auto col = col_out_begin + read_count;
+            const auto row = row_out_begin + read_count;
+
+            values_accessor[actual_nnz] = diagonals_accessor[d_i][col];
+            indices_accessor[0][actual_nnz] = row;
+            indices_accessor[1][actual_nnz] = col;
+            ++actual_nnz;
+          }
+        }
+      });
+  indices = indices.narrow(1, 0, actual_nnz);
+  values = values.narrow(0, 0, actual_nnz);
+  SparseTensor sparse =
+      at::sparse_coo_tensor(indices, values, shape, sparse_options);
+  return sparse;
+}
+
+// Check offsets for duplicates, and out-of-bounds diagonals
+void validate_spdiags_offsets_cpu(
+    const Tensor& offsets,
+    const int64_t n_row,
+    const int64_t n_col) {
+  std::set<int64_t> seen_offsets;
+  auto offsets_accessor = offsets.accessor<int64_t, 1>();
+  for (auto i : c10::irange(offsets.size(0))) {
+    auto off = offsets_accessor[i];
+    TORCH_CHECK(
+        seen_offsets.insert(off).second,
+        "Offset array contains duplicate values");
+    TORCH_CHECK(
+        ((-1 * n_row) < off) && (off < n_col),
+        "Diagonal ",
+        off,
+        " does not exist in output shape (",
+        n_row,
+        ",",
+        n_col,
+        ")");
```

Comment:
Could you please update it to "spdiags() : Diagonal ... "? Could you also provide a valid range for offsets in this message?